pkg/controller/dynamicupgradeableconditionscontroller.go

```diff
@@ -0,0 +1,239 @@
+		return nil, err
+	}
+
+	store := c.buildHelmStore(c.kubeclient.CoreV1().Secrets("openshift-operator-controller"))
+
+	// Get all ClusterExtensions incompatible with next Y-stream
+	for _, obj := range ceList {
+		ce, ok := obj.(*unstructured.Unstructured)
+		if !ok {
+			c.logger.V(4).Error(nil, fmt.Sprintf("Unable to Cast Object to unstructured.Unstructured. Object is of type %T", obj))
+			return nil, nil
+		}
+
+		rel, err := store.Deployed(ce.GetName())
+		if errors.Is(err, driver.ErrNoDeployedReleases) {
+			c.logger.Info("Cluster Extension not yet deployed - will check again later")
+			continue
+		}
+		if err != nil {
+			c.logger.V(4).Error(err, "Error returning the last deployed release")
+			return nil, err
+		}
+		props, err := propertyListFromPropertiesAnnotation(rel.Chart.Metadata.Annotations["olm.properties"])
+		if err != nil {
+			c.logger.V(4).Error(err, "Error converting olm.properties")
+			return nil, err
+		}
+		for _, p := range props {
+			if p.Type == MaxOpenShiftVersionProperty {
+				version := string(p.Value)
+				maxOCPVersion, err := toSemver(version)
+				if err != nil {
+					c.logger.V(4).Info(fmt.Sprintf("Error converting to semver for version %s: %v", version, err))
+					continue
+				}
+				if maxOCPVersion != nil && !maxOCPVersion.GTE(next) {
+					// Incompatible
+					incompatibleOperators = append(incompatibleOperators, rel.Labels[BundleNameKey])
+				}
+			}
+		}
+	}
+
+	// deterministic ordering
+	sort.Strings(incompatibleOperators)
+
+	updateStatusFuncs := make([]v1helpers.UpdateStatusFunc, 0, len(incompatibleOperators))
+	if len(incompatibleOperators) > 0 {
+		updateStatusFuncs = append(updateStatusFuncs, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "Upgradeable",
+			Status:  operatorv1.ConditionFalse,
+			Reason:  reasonIncompatibleOperatorsInstalled,
+			Message: strings.Join(incompatibleOperators, ","),
+		}))
+	} else {
+		updateStatusFuncs = append(updateStatusFuncs, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "Upgradeable",
+			Status: operatorv1.ConditionTrue,
+		}))
+	}
+	return updateStatusFuncs, nil
+}
+
+func propertyListFromPropertiesAnnotation(raw string) ([]property.Property, error) {
+	var props []property.Property
+	if err := json.Unmarshal([]byte(raw), &props); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal properties annotation: %w", err)
+	}
+	return props, nil
+}
+
+func (c *dynamicUpgradeableConditionController) buildHelmStore(secretClient v1.SecretInterface) helm.Storage {
+	log := func(s string, args ...interface{}) { c.logger.V(4).Info(fmt.Sprintf(s, args...)) }
+	csConfig := storage.ChunkedSecretsConfig{Log: log}
+
+	return helm.Storage{
+		Driver: storage.NewChunkedSecrets(secretClient, "operator-controller", csConfig),
+		Log:    log,
+	}
+}
+
+const (
+	releaseEnvVar = "RELEASE_VERSION" // OpenShift's env variable for defining the current release
```

Comment:
re: my earlier comment about moving the lookup out to runOperator, I'm not sure we need this constant at all if we're just doing the lookup in one place. Can probably hardcode os.Getenv("RELEASE_VERSION").